infrastructure/server/ec2_instances.tf

```diff
@@ -0,0 +1,80 @@
+locals {
+    fam_util_ec2_instance_profile_name_prefix = "fam_util_ec2_instance"
+}
+
+variable "fam_util_ec2_instance_ami" {
+  description = "Instance image for FAM Util EC2"
+  type        = string
+  # Amazon Linux 2 Kernel 5.10 AMI 2.0.20230119.1 x86_64 HVM gp2
+  default     = "ami-092e716d46cd65cac"
+}
+
+variable "fam_util_ec2_instance_type" {
+  description = "Instance type to use for the instance"
+  type        = string
+  default     = "t2.micro"
+}
+
+resource "aws_instance" "fam_util_ec2_instance" {
+  ami                     = "${var.fam_util_ec2_instance_ami}"
+  instance_type           = "${var.fam_util_ec2_instance_type}"
+  subnet_id = data.aws_subnet.a_app.id
+  vpc_security_group_ids = ["${aws_security_group.fam_app_sg.id}"]
+
+  depends_on = [aws_security_group.fam_app_sg]
+
+  iam_instance_profile = "${aws_iam_instance_profile.fam_util_ec2_instance_profile.id}"
+
+  root_block_device {
+    # At default, this is not set.
+    encrypted = true
+  }
+
+  tags = {
+      Name = "fam_util_ec2_host"
+      managed-by = "terraform"
+  }
+
+  # Script to install postgresql.
+  user_data = <<EOF
+  #!/bin/bash
+  echo "Installing postgresql.x86_64" > init.log
+  sudo yum update -y >> init.log 2>&1 &
+  sudo yum install -y postgresql.x86_64 >> init.log 2>&1 &
+  echo "Postgres installation done" >> init.log
+
+  EOF
+
+}
+
+resource "aws_ec2_instance_state" "fam_util_ec2_instance_state" {
+  instance_id = aws_instance.fam_util_ec2_instance.id
+  # fam_util_ec2_instance instance created in dormant state.
+  state       = "stopped"
+}
+
+resource "aws_iam_instance_profile" "fam_util_ec2_instance_profile" {
+  name = "${local.fam_util_ec2_instance_profile_name_prefix}_instance_profile"
+  role = "${aws_iam_role.fam_util_ec2_instance_role.name}"
+}
+
+resource "aws_iam_role" "fam_util_ec2_instance_role" {
+  name = "${local.fam_util_ec2_instance_profile_name_prefix}_role"
+
+  assume_role_policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": "sts:AssumeRole",
+      "Principal": {
+        "Service": "ec2.amazonaws.com"
+      },
+      "Effect": "Allow",
+      "Sid": ""
```

Comment:
Not sure if this is too restrictive. I can check in CheckPoint whether anything was flagged.

Reply:
As suspected, CheckPoint is complaining. Here's the first finding:
Title
Ensure that custom IAM Role doesn't have an overly permissive scope (Contains a wildcard)
[(D9.AWS.IAM.108)](https://gsl.dome9.com/D9.AWS.IAM.108.html)
Description
To reduce the risk of a misuse or abuse due an overly privileged IAM Role, minimize the scope your IAM Role is allowed to perform actions on, according to the principal of least privilege.
Ruleset
[AWS CloudGuard Best Practices](https://ca.portal.checkpoint.com/Dashboard/CloudGuard#/v2/compliance-engine/policy/-5)
Remediation

Note: The provided rule covers managed IAM Roles only: AWS service-roles and AWS service-role policies are excluded. In addition, this rule ignores any policy document block with one or more actions from the following action groups: sts, ssm, cloudwatch, ec2messages, ssmmessages, logs, route53.

From Portal

1. Sign in to the AWS Management Console and open the AWS IAM console at https://console.aws.amazon.com/iamv2/
2. From the left pane, under 'Access management' select 'Roles'
3. Identify and select the relevant IAM Role
4. Edit its 'Permissions policies' according to the principal of least privilege

Reply:
Here's a suggested solution from CheckPoint to revise permissions granted to IamRole:fam_util_ec2_instance_role- I haven't tried to evaluate it:
{
   "Version": "2012-10-17",
   "Statement": [
      {
         "Action": [
            "ssmmessages:CreateControlChannel",
            "ssmmessages:CreateDataChannel",
            "ssmmessages:OpenControlChannel",
            "ssmmessages:OpenDataChannel"
         ],
         "Resource": [
            "*"
         ],
         "Effect": "Allow",
         "Sid": "CloudGuardGenerated0"
      },
      {
         "Action": [
            "ec2messages:AcknowledgeMessage",
            "ec2messages:DeleteMessage",
            "ec2messages:FailMessage",
            "ec2messages:GetEndpoint",
            "ec2messages:GetMessages",
            "ec2messages:SendReply"
         ],
         "Resource": [
            "*"
         ],
         "Effect": "Allow",
         "Sid": "CloudGuardGenerated1"
      },
      {
         "Action": [
            "ssm:ListInstanceAssociations"
         ],
         "Resource": [
            "arn:aws:ec2:ca-central-1:377481750915:instance/i-038d04d1682e1235b",
            "arn:aws:ec2:ca-central-1:377481750915:instance/i-0975b847cd35f0e90"
         ],
         "Effect": "Allow",
         "Sid": "CloudGuardGenerated2"
      },
      {
         "Action": [
            "ssm:UpdateInstanceInformation"
         ],
         "Resource": [
            "arn:aws:ec2:ca-central-1:377481750915:instance/*",
            "arn:aws:ssm:ca-central-1:377481750915:managed-instance/*"
         ],
         "Effect": "Allow",
         "Sid": "CloudGuardGenerated3"
      }
   ]
}

Reply:
Thanks Basil!
Since this terraform change hasn't been deployed yet to higher environment, I see the TEST environment ec2 instance has this role (`**EC2-Default-SSM-AD-Role**`) associated with it.
![image](https://github.com/bcgov/nr-forests-access-management/assets/81595625/20fbac5f-23f9-4061-8a60-bf4c34df52c2)
Maybe I should associate terraform created instance with this role as well instead of adding confusing policies...

Reply:
Hi @basilv I have replaced terraform to associate with this role, same as previous instance created manually.
Hope the CheckPoint give no warning.